### PR TITLE
Set play.allowGlobalApplication=false by default

### LINKED ISSUE #8050: play.allowGlobalApplication should default to false
We should default to not allowing the global application instance. We probably should keep a few tests around that use it to make sure it still works for now.

We also should document a migration path for users that are still relying on static state. Usually static injection is a good temporary fix.

### PR DESCRIPTION
**Pull request by gmethvin**
This sets `play.allowGlobalApplication = false` by default and updates the tests accordingly.

I also realized there are several places in our documentation where we use global state where we shouldn't be. Perhaps also it would be smart to remove the `Action` companion object to prevent conflicts with the one in `BaseController` and confusion about which `Action` you are using.

Fixes #8050

### CODE REVIEW

**Comment by gmethvin** [comment]
Once we merge this I think the next step is to remove the deprecated `Action` companion object.

File: documentation/manual/working/scalaGuide/main/http/code/ScalaActionsComposition.scala [hunk 1]
```diff
@@ -132,15 +129,17 @@ class ScalaActionsCompositionSpec extends Specification with ControllerHelpers {
       //#actions-def-wrapping
       import play.api.mvc._
 
-      def logging[A](action: Action[A])= Action.async(action.parser) { request =>
+      // Note: to use the default action builder in a controller you can use Action.
```
Comment (marcospereira) [inline documentation/manual/working/scalaGuide/main/http/code/ScalaActionsComposition.scala]: Maybe keep using `Action` here since it is the idiom we use in controllers? And since we have `parse` too, keep `Action.async(parse.some)` idiom?

File: documentation/manual/working/scalaGuide/main/tests/code/specs2/ExampleEssentialActionSpec.scala [hunk 2]
```diff
@@ -13,8 +13,11 @@ class ExampleEssentialActionSpec extends PlaySpecification {
 
   "An essential action" should {
     "can parse a JSON body" in new WithApplication() {
-      val action: EssentialAction = Action { request =>
-        val value = (request.body.asJson.get \ "field").as[String]
+      val Action = app.injector.instanceOf[DefaultActionBuilder]
```
Comment (marcospereira) [inline documentation/manual/working/scalaGuide/main/tests/code/specs2/ExampleEssentialActionSpec.scala]: Use `Injecting`.

File: documentation/manual/working/scalaGuide/main/http/code/ScalaActionsComposition.scala (hunk 1 above)
Comment (gmethvin) [reply documentation/manual/working/scalaGuide/main/http/code/ScalaActionsComposition.scala]: I agree about `Action` since in this case I can't think of any reason to use a custom action builder here anyway. My code here was a bit of a lazy shortcut because `play.api.mvc._` still imports `Action`.

I still need to pass through the `parser`, or else you can't pass actions with a custom body type `A`.

Actually, a more direct way of doing this would be:
```scala
def logging[A](action: Action[A]) = {
  val builder = new ActionBuilderImpl(action.parser)(action.executionContext)
  builder.async { request =>
    Logger.info("Calling action")
    action(request)
  }
}
```
But the current API unfortunately is not very elegant. I wonder if it would make sense to add a `builder` method to the `Action` to get a builder based on that action, so this would be as simple as:

```scala
def logging[A](action: Action[A]) = action.builder.async { request =>
  Logger.info("Calling action")
  action(request)
}
```

File: framework/src/play-test/src/main/scala/play/api/test/Helpers.scala [hunk 3]
```diff
@@ -39,11 +39,9 @@ trait PlayRunners extends HttpVerbs {
   val FIREFOX = classOf[FirefoxDriver]
 
   /**
-   * Returns `true` if this application needs to run sequentially, rather than in parallel with other applications.
-   * Typically that means the application and/or test relies on some global state, so this defaults to the
-   * globalApplicationEnabled setting. This method is provided so the behavior can be customized if needed.
+   * Even though we don't share any global state, we have to run sequentially since we share a test server port.
    */
-  protected def shouldRunSequentially(app: Application): Boolean = app.globalApplicationEnabled
+  protected def shouldRunSequentially(app: Application): Boolean = true
```
Comment (gmethvin) [inline framework/src/play-test/src/main/scala/play/api/test/Helpers.scala]: I probably should fix this.

Comment (gmethvin) [reply framework/src/play-test/src/main/scala/play/api/test/Helpers.scala]: We probably should avoid using the same port for all servers but that change seems more complicated and out of scope for this PR.

**Comment by gmethvin** [comment]
@marcospereira Changed the examples to use Action.

**Comment by gmethvin** [comment]
@marcospereira I just rebased this since there were some conflicts.

**Comment by mkurz** [comment]
See #8360

File: documentation/manual/releases/release27/migration27/Migration27.md [hunk 4]
```diff
@@ -3,9 +3,15 @@
 
 This is a guide for migrating from Play 2.6 to Play 2.7. If you need to migrate from an earlier version of Play then you must first follow the [[Play 2.6 Migration Guide|Migration26]].
 
+### `play.allowGlobalApplication` defaults to `false`
+
+`play.allowGlobalApplication = false` is set by default in Play 2.7.0. This means `Play.current` will not work. You can set this to `true` to make `Play.current` and other deprecated static helpers work again, but be aware that this feature will be removed in Play 2.8.
```
Comment (schmitch) [inline documentation/manual/releases/release27/migration27/Migration27.md]: do we call the next version 2.8?

**Comment by schmitch** [comment]
Shouldn't we also change the behavior of `NettyServer.fromRouter` and `AkkaHttpServer.fromRouter` to use global state? (and deprecate them?)

File: documentation/manual/releases/release27/migration27/Migration27.md (hunk 4 above)
Comment (wsargent) [reply documentation/manual/releases/release27/migration27/Migration27.md]: Rather than saying "will not work", we should specify that Play.current will throw an exception.

**Comment by schmitch** [comment]
Well if we remove `Action` the whole `withRouter` function is useless because you can't use it without spinning up a second `ActorSystem`. I mean you can't get the `ActorSystem` component from `withRouter`.

Well you can use `ActionBuilder.IgnoringBody()` however even that would at least need some kind `ExecutionContext` or even use `EssentialAction` and built your own BodyParser, but I do not think that is useful.

So that brings us to the point that `withRouter` would be broken, it can't be used because it would throw an Exception if we would keep `Action` and if we won't you can't use it easily without building a ton of stuff yourself. (Except for `GET` request) Or by spinning up a second `ActorSystem`, because `PlayBodyParsers.apply` at least needs a `Materializer`.

So basically `withRouter` should be deprecated (and we need to keep `Action` and set the global state to true when calling withRouter) or removed (which we can't because we didn't deprecate it)

**Comment by gmethvin** [comment]
I'm not a fan of `withRouter` either. There is really no way to effectively use it without global state. The best alternative is probably to construct a components and do stuff with compile-time DI. Not sure whether it's in scope for this PR though.

**Comment by schmitch** [comment]
maybe for this pr we should just deprecate `withRouter` and tell people to use `fromRouterWithComponents` and probably have another issue, where we can talk about what we further need to do, etc.

**Comment by gmethvin** [comment]
Okay @schmitch sounds good. I deprecated `fromRouter` and I also added a `DefaultAkkaHttpServerComponents`, which is equivalent to `AkkaHttpServerComponents with BuiltInComponents with NoHttpFiltersComponents`, to make servers easier to construct from components.

**Comment by richdougherty** [comment]
:fireworks: :confetti_ball:
